sql/src/main/java/io/crate/analyze/Rewriter.java

```diff
@@ -0,0 +1,248 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Sets;
+import io.crate.analyze.relations.JoinPair;
+import io.crate.analyze.relations.QuerySplitter;
+import io.crate.analyze.symbol.*;
+import io.crate.metadata.Functions;
+import io.crate.metadata.NestedReferenceResolver;
+import io.crate.metadata.RowGranularity;
+import io.crate.operation.operator.AndOperator;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.sql.tree.QualifiedName;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+
+import javax.annotation.Nullable;
+import java.util.*;
+
+@Singleton
+public class Rewriter {
+
+    private final EvaluatingNormalizer normalizer;
+
+    @Inject
+    public Rewriter(Functions functions, NestedReferenceResolver globalRefResolver) {
+        normalizer = new EvaluatingNormalizer(
+            functions,
+            RowGranularity.CLUSTER,
+            globalRefResolver,
+            null,
+            false
+        );
+    }
+
+
+    /**
+     * Rewrite an Outer join to an inner join if possible.
+     *
+     * Conditions on OUTER tables are not pushed down when a MultiSourceSelect is initially created because
+     * the whereClause needs to be checked AFTER the join
+     * (because if the join generates NULL rows the whereClause could become TRUE on those NULL rows)
+     *
+     * See the following two examples where <code>t2</code> is the OUTER table:
+     *
+     * <pre>
+     *     select * from t1
+     *          left join t2 on t1.t2_id = t2.id
+     *     where
+     *          coalesce(t2.x, 20) > 10   # becomes TRUE for null rows
+     * </pre>
+     *
+     *
+     *
+     * But if we know that the whereClause cannot possible become TRUE then we can push it down
+     * (re-writing this into an inner join)
+     *
+     * This is possible because all rows that are generated by the left-join would be removed again by the whereClause anyway.
+     *
+     * <pre>
+     *     select * from t1
+     *          left join t2 on t1.t2_id = t2.id
+     *     where
+     *          t2.x > 10   # if t2.x is NULL this is always FALSE
+     * </pre>
+     */
+    public void tryRewriteOuterToInnerJoin(final JoinPair joinPair,
```

Comment:
Guess it is because I had an anonymous class inside before and the joinPair was used there - can change it if you want